### DISCUSSION TITLE
Can someone help me explain?

### DISCUSSION THREAD

**Discussion by waykyi** [fontforge/fontforge#4802] [Q&A]
I am confused in lookup REORDER_CHAIN. Why two lookups are between two glyphs that need to be replaced. Is this the format? Are the lookups executed by the order they appeared. Thanks

lookup REMOVE_CAKRA {
    sub ka ka.pas_cakra.ns by ka;
} REMOVE_CAKRA;

lookup REORDER_CAKRA {
    sub ka by ka.pas_cakra ka;
} REORDER_CAKRA;

lookup REORDER_CHAIN {
    sub ka' lookup REMOVE_CAKRA lookup REORDER_CAKRA ka.pas_cakra.ns' ;
} REORDER_CHAIN;

**Comment by ctrlcctrlv** [comment]
Those lookups wouldn't be executed at all, because they aren't part of a feature.

The order of execution of OpenType lookups is an _extremely complicated_ topic. It is not guaranteeable in my view. All that can be guaranteed is that _within the same feature_, _lookups are applied in descending order_.

To eliminate confusion, you can put a single `substitute` in each `lookup`. I recommend this unless size is preventing you from doing it; otherwise, you can indeed put multiple `substitute`'s in a single lookup subtable. Read the spec to learn the execution order.

The below image details _feature order_. It is by @simoncozens and is not my work. It answers the question “when two OpenType features are enabled, in which order are they executed?”:

![image](https://user-images.githubusercontent.com/838783/132135820-295cc734-090e-45fe-a290-04915166bf17.png)

**Reply by waykyi** [reply]
This was directly copied from Opentype Feature specs. There was explanation in the specs, but I just need more info about chaining lookups and putting them between two glyphs that need to be replaced. I like to know how and why they did this way. Example is from the specs, I assumed it must be working? And thank you very much for the picture and how Features flow. So, above lookups are just reordering the glyphs, which I will need to put in my font. But I need it further replacements based on more inputs from the user. That is why I am very interested in understanding clearly on these routine lookups. Thanks

**Reply by ctrlcctrlv** [reply]
That's not fair. :wink: You can't just copy examples from the Adobe–OpenType FEA standard uncited and expect me to recognize them. I know the standard well but not that well, I don't have a photographic memory.

You're referring to example 5 in §5.f. which deals with GSUB LookupType 6.

The text above the example code implies that the lookup `REORDER_CHAIN` will be added to some feature and then triggered when that feature's active. So, yes, feature tag doesn't matter.

We're looking at a chaining substitution. The chain is meant to do a swap.

The order of operations is therefore, assuming e.g.:

```fea
feature liga {
  lookup REORDER_CHAIN;
} liga;
```

and the input glyphstream `ka ka.pas_cakra.ns`:

```fea
ka ka.pas_cakra.ns   # Begin…
[ka] ka.pas_cakra.ns # Active glyph is /ka, lookahead matches, execute REORDER_CHAIN (GSUB type 6):
[ka]                 # Execute lookup REMOVE_CAKRA (GSUB type 2), and immediately…
ka.pas_cakra [ka]    # Execute lookup REORDER_CAKRA (GSUB type 2).
ka.pas_cakra [ka]    # At end of stream, no more lookups match, done.
```

So why did this work? Because the invocation of REORDER_CAKRA was already guaranteed, it did not matter that when we only had `ka` in the glyph stream mid-execution the sub rule no longer matched. The dominoes had already started falling—the match will not be rechecked until all chaining lookups are executed.

Glyph reordering is not simple in OpenType. This is the best cross-platform way to execute a swap. Remember, you cannot write:

```fea
sub ka ka.pas_cakra.ns by ka.pas_cakra ka;
```

That's not a valid multiple substitution, because a valid multiple substitution must always replace exactly one glyph with 𝑛 glyphs.

Now, this isn't in the FEA documentation anywhere, but technically, as @simoncozens taught me, a Multiple Substitution can have a NULL replacement set. So example 5 could theoretically be written as:

```fea
lookup NULL_CAKRA {
  sub ka.pas_cakra.ns by NULL;
} NULL_CAKRA;

lookup EXPAND_CAKRA {
  sub ka by ka.pas_cakra ka;
} EXPAND_CAKRA;

lookup Example5 {
  sub ka' lookup EXPAND_CAKRA ka.pas_cakra.ns' lookup NULL_CAKRA;
} Example5;
```

I personally added NULL support to FontForge in #4619. I think this example is easier to understand what's going on. We still achieve a swap, just via deletion instead, in a way some developers might find easier to reason about.

I guess what I'm trying to say @waykyi is that this stuff is all really complex and you're not going to be able to just copy and paste examples without trying to understand the logic of the OpenType spec.

**Reply by waykyi** [reply]
My sincerely apology to you. Sorry, I should have mentioned that clearly in the first post in which I was lost in the syntaxes. I am very grateful for your patience and continue to explain me so well in here. Thank you very much for writing up almost the whole thing and explaining how you did in each substitute and why. It is very nice of you and thank you.

**Reply by ctrlcctrlv** [reply]
No problem, good refresher for me. :-)
